src/posts/en/javascript-iterables-and-iterators.md

```diff
@@ -0,0 +1,294 @@
+
+### So what exactly is this Iterable interface you speak of?
+
+It is simple, really. Any object that contains a `[Symbol.iterator]` method is an iterable object.<br>
+In case you are unfamilar with symbols, they are, basically, objects that can be used as property keys.<br>
+`Symbol.iterator` is a *well-known* (baked in the language) Symbol, whose primary usage is defining and consuming iterable objects.
+
+In the previous examples, we were implicitly making use of the `[Symbol.iterator]` method inherited from the given objects' prototypes. Several built-ins such as Array, Set, String and Map define a default iteration behavior, while others such as Object do not.
+
+As previously mentioned, the Iterable interface allows defining and customizing objects' iteration order. For a quick example, we can borrow arrays' default iteration behavior (`Array.prototype[Symbol.iterator]`) to make array-like objects iterable:
+
+```javascript
+let iterable = {
+	0: 'a',
+	1: 'b',
+	2: 'c',
+	length: 3,
+	[Symbol.iterator]: Array.prototype[Symbol.iterator]
+};
+for (let item of iterable) {
+	console.log(item); // 'a', 'b', 'c'
+}
+```
+
+Now you may be wondering: **"How do I implement custom iteration behavior?"**<br>
+As you already know, adding a `[Symbol.iterator]` method to a given object makes it iterable, but here is the catch: the `[Symbol.iterator]` method must return an *iterator object*, which is actually responsible for the iteration logic. We will cover this in the next section.
+
+Before we move on to iterators, let's recap what the Iterable interface is about:
+
+- It signals that a given object is iterable to consumers (iterable object: "Hey! I have a `[Symbol.iterator]` method.");
+- It provides a standardized way to iterate any iterable object (iterable object: "Call my `[Symbol.iterator]` method if you want an iterator that implements my default iteration logic").
+
+The points above may seem slightly redundant if you are familiar with interfaces (Java, C#), but it is worth being clear as this is a new addition to the JavaScript language.
+
+Note that several built-in functions and language constructs that expect a sequence of values accept iterable objects. For instance, `Promise.all()` accepts any iterable object as argument, not just plain arrays. The same holds true for destructuring assignments (`[a, b] = iterable`) and `yield* iterable`.
+
+## Iterators
+
+Iterator objects are objects that implement the [Iterator interface](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-iterator-interface). That is, iterator objects must have a `next` method that returns a result object in the `{ value: Any, done: Boolean }` format. The first call to `.next()` returns the result of the first iteration (e.g. the 0th item in an array). The `done` property signals when the iterator has been exhausted and no more values are available.
```

Comment:
I tried to keep it simple, but "the 0th item" does not sound right. Perhaps "the item at the 0th index"